### PR TITLE
Gnome 40 Support and generic cpufreq frequency scaling driver support

### PR DESCRIPTION
**Pull request by fin-ger**
This PR is arguably not only focusing on just getting Gnome 40 support done, but also includes a rework of the `cpufreqctl` script and code cleanups. This is mostly due to me just hacking this in my free time and not caring about 100% clean git workflows. Below is a brief explanation on what I changed and why I did this.

# The new `cpufreqctl` script

This PR includes a major rework of the `cpufreqctl` to allow the inclusion of other frequency scaling drivers other than `intel_pstate`. The script itself ended up as a **huge** shell-script and I sincerely apologies for that! Further, the script now possesses a `json` output mode for (hopefully) easy integration into environments other than Gnome (e.g. sway bars, etc.). The `json` mode is also used by the extension and the human-readable output is only used for debugging purposes by myself. I restructured the cli a bit to make it a bit more intuitive and hopefully made the first steps towards the frequently requested "port" of this extension to other desktop environments. However, I do not plan to support any other DE but Gnome. The changes to the `cpufreqctl` script are done in the hopes for someone else to pick it up and build a frontend for their favorite DE.

## Why is the New `cpufreqctl` Script So Huge?

Most of the code belongs to command line interface parsing and is mostly copy&paste code. I thought about the following solutions to this:

 1. Don't use shell-scripts for this, but other languages which make cli-parsing concise and bulletproof
     - languages other than GJS (such as Python) are not guaranteed to be available on the user's system
     - GJS itself does not provide great help in parsing command line arguments
     - native compiling languages like Rust, Go, etc. which provide great tools for parsing cli's possess the problem of requiring binaries targeted for specific architectures
         - musl vs. glibc can be avoided with static linkage, but may introduce security risks as dependencies are all bundled in the executable and won't update security issues with the systems package manager -> must be managed by our extension with frequent updates of the binaries, for which I don't have to time for
         - other CPU architectures (e.g. ARM) are supported by the `cpufreq` scaling driver, so multiple binaries for several CPU architectures would be needed -> too much management overhead
 2. Using fancy bash/sh magic with argument expansion and templated functions
    - As this script will run with root privileges, I don't think it's a good idea to use error-prone argument-expansion string interpolation techniques in bash as this can easily be exploited when a single tick is missing or a variable is not checked for validity

As the most simple solution, I chose to run `shellcheck -o all` frequently and fixed **all** issues (even minor ones, I didn't want to risk anything) and copy&pasted shell argument parsing code chunks to avoid argument/parameter-expansion.

If we can e.g. safely assume Python 3 to be available on our user's systems the `cpufreqctl` script can safely be replaced by a Python version in the future.

## Modular Backend For the Frequency Scaling Driver

The `cpufreqctl` script can be easily extended with additional backends. For that 10 functions need to be implemented:

```
<backend-name>_supported() {
}

<backend-name>_reset() {
}

<backend-name>_turbo_get() {
}

<backend-name>_turbo_set() {
}

<backend-name>_min_get() {
}

<backend-name>_min_set() {
}

<backend-name>_max_get() {
}

<backend-name>_max_set() {
}

<backend-name>_info_frequencies() {
}

<backend-name>_info_current() {
}
```

Replace `<backend-name>` with the name of your backend. Look at the [existing backends](https://github.com/martin31821/cpupower/blob/b704c8c7f77190406d6e09c658b7e4864cfeffb7/tool/cpufreqctl#L296) for examples.

I implemented 3 backends:

 - `fake` which is a fake backend for development purposes which is not available when `PRODUCTION=yes` is set
 - `intel_pstate` which is copy&paste of the old control code
 - `cpufreq` which is a new backend using the `cpufreq` general purpose cpu frequency scaling driver

Additionally, there is an automatic backend selection where `intel_pstate` is preferred over `cpufreq`.

The backend is chosen via the `--backend` command line flag.

# The Preferences Window

The preferences window received a minor rework to accommodate the new options available in the new `cpufreqctl` script:

 - A frequency sampling mode selection (min, max, avg, random)
 - The `cpufreqctl` backend (automatic, fake, intel_pstate, cpufreq)

## New Look

I implemented a preferences list similar to the ones seen in Gnome's System Settings. Further, the `About` button is now placed in the headerbar.

## Gnome 40 Support

As Gnome 40 **forces** the usage of Gtk4 in the preferences dialog for no apparent reason, I decided to create my own preferences window with Gtk3 under Gnome 40 with some minimal polyfilling to ensure compatibility with the preferences windows of previous Gnome versions.

The reasons why I went this path:

 - Glade does not support Gtk4 yet and I don't have the time to dig into the XML syntax of .ui/.glade files
     - I tried it and wasted hours on finding the allowed tags and values for each individual element and just got a (non-functional) preferences window up and running after a full day
 - The preferences window will receive a rework when Libadwaita is available for Gnome Shell extensions anyway (using `AdwPreferencesWindow`)
 - Compatibility for old Gnome versions is very important as we must be able to provide security updates regarding our polkit rules and `cpufreqctl` script
 - Usability is not impacted by this decisions -> Gnome 40 ships a lot of GTK3 apps still, so there won't be any noticeable issues other than slight theming changes

# Coding Style and ESlint

I copied the eslint rules from the gnome-shell repo and adjusted them to create mostly warnings and only errors if the code actually contains errors. The eslint rules require a plugin to work, you have to install it from the `package-lock.json` with `npm` if you like to use the eslint rules. I did not include eslint into the makefile or any other process, it's just a preference you can choose to use or ignore.

Along with the eslint config, I fixed all issues in the code and found some unnoticed bugs (such as the overdrive-border-color being accessed in the wrong scope, etc.). That's also the reason why this PR is so huge and has so many changed lines.

# Changes in the Update Procedure

The `cpufreqctl` script is now updated when the output of `cpufreqctl --version` does not match `${VERSION}` in `installer.sh`.

# TODO

 - [x] check if translation template is up-to-date
 - [x] Change supported Gnome version in Readme
 - [x] Adjust Readme to reflect support for CPU's other than Intel
 - [x] cpufreq backend boost on intel CPUs not working
 - [x] edited profile is lost when moving position
 - [x] dbus behaves weird, I don't really know what the issue is but on some systems uninstall freezes reproducibly
 - [x] test on major distributions
     - [x] Arch (Gnome 40)
     - [x] Fedora 33 (Gnome 3.38)
     - [x] Fedora 34 (Gnome 40)
     - [x] Ubuntu 18.04 (Gnome 3.28)
     - [x] Ubuntu 20.04 (Gnome 3.36)
     - [x] Ubuntu 21.04 (Gnome 3.38)
     - [x] Debian stable (Gnome 3.30)
     - [x] Debian testing (Gnome 3.38)

 - Uninstall seems unstable on slow systems, where `updateFreq` is called by the glib timeout during uninstallation (race condition, not dependent on Gnome platform)
 - Couldn't test update procedure yet, as I only tried it in a VM where `intel_pstate` is not supported so installation of an older version is not possible without some "hacks" (Thank god this is the last time I have to tinker with this, praise the fake backend)

## Test procedure

Run

```console
$ make install PREFIX=$HOME/.local
$ dbus-run-session -- gnome-shell --nested --wayland
```

and check for errors and warnings regarding `cpupower` in the console output while doing the following steps:

 1. Do clean installation
 2. Moving both sliders in the dropdown with
     1. Mouse
     2. Arrow keys
     3. Mouse wheel
 3. Toggle Turbo Boost
 4. Change Profile
 5. Toggle Auto-Switch (real testing not possible as VMs don't fire AC adapter events)
 6. Open Preferences and toggle all options
 7. Do 2-4 for all available backends
 8. Create a new profile and change its position (move up and down in the list)
 9. Set all values in that new profile
 10. Select it in the extensions dropdown menu
 11. Remove the profile
 12. Open About Window
 13. Uninstall tool and polkit rules
 14. Install latest version from Gnome Extensions
 15. Update with this version
 16. Check if update procedure is working (Note: this requires a version increment which is only done by `make release` which I have not done yet, so version in `installer.sh` and `cpufreqctl` needs to be incremented manually for testing)

Global installation method and all other installation methods are not tested (at least by me). You can do it if you want and if you encounter any issues, I will do my best to fix them.

Orca (screen reader) support can only be tested in "real" gnome-shells (non-nested). At least I couldn't get Orca running in the nested gnome-shell.

# Known Issues

 - won't run on
     - Fedora Silverblue
     - GnomeOS
 - not using Gtk4 but a hack to use Gtk3 under Gnome 40 to ensure compatibility with all other distros but Arch & Fedora 34

# Other Things

I'm probably missing a ton of things here. If you have any questions, please don't hesitate to ask. Let's see if I can remember what I did :smile:

### CODE REVIEW

**Comment by fin-ger** [comment]
@mastercaution can you review or are you lacking privileges?

**Comment by mastercaution** [comment]
@fin-ger Thanks for your work, you are doing a fantastic job!

I can do a review with my privileges, but I have to find some time to do it ... hopefully this week :smile:

**Comment by fin-ger** [comment]
> I can do a review with my privileges, but I have to find some time to do it ... hopefully this week smile

Ah very nice! I'd love to have someone else skim through the code before going on master. But if your time doesn't allow it until next week, don't push it too hard.

Realistically, the release will be made in June as translators will need to step in and there are #149, #141, #140, and #87 which might be worth for inclusion in the next release.

**Comment by fin-ger** [comment]
Forget to mention: I quickly summarized the test procedure I follow for testing the extension out on the various platforms. Just in case if anyone else (maybe from #145) wants to step in and help out testing this. No need to follow this 100%, it's just a hint to cover ~80% of functionality.

**Comment by fin-ger** [comment]
>  - Discard button has to be clicked twice when editing a profile
>  - Uninstall seems unstable on slow systems, where `updateFreq` is called by the glib timeout during uninstallation (race condition, not dependent on Gnome platform)
>  - Couldn't test update procedure yet, as I only tried it in a VM where `intel_pstate` is not supported so installation of an older version is not possible without some "hacks" (Thank god this is the last time I have to tinker with this, praise the fake backend)

These issues currently remain in all distributions I tested. Other issues should be fixed by now. Please report if other issues still exist.

**Comment by fin-ger** [comment]
Ok, I figured out what makes the uninstall process freeze:

`gnome-extensions disable cpupower@mko-sl.de` is called after the uninstallation was successful to disable and then enable the extension again to recreate the extensions menu.

The `gnome-extensions` command block for ~30 seconds and exits with code 512. On stderr `Failed to connect to GNOME Shell` is printed.

I don't know why this is happening and how to solve it, but the `gnome-extensions` command is not responsive. Will look into this another day, maybe someone else has an idea on this?

**Comment by fin-ger** [comment]
I figured out that the `DBUS_SESSION_BUS_ADDRESS` environment variable is doing the trouble for the freezing uninstallation. The environment variable seems to be different inside the extensions preferences window (possibly altered by GNOME?).

When doing a `disable` in GJS with a clean environment and the `DBUS_SESSION_BUS_ADDRESS` environment variable manually set, the call is always successful.

```
GLib.spawn_sync("/home/fin", ["gnome-extensions", "disable", "cpupower@mko-sl.de"], ["DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus"], GLib.SpawnFlags.SEARCH_PATH, null)
```

Investigating on the content of this variable within the GNOME shell extension preferences window reveals that the variable is set to e.g. `unix:abstract=/tmp/dbus-0kwvtEXyDG,guid=a124133f1c054d0a5218a40f60cb8060`. This socket seems to simply not work. I suppose this is a bug in Gnome Shell? I am not really sure what to do here. I can guess the correct dbus socket, but I don't think this is the proper way to address this. I guess I will have to get in contact with the Gnome developers to solve this...

**Comment by fin-ger** [comment]
TL;DR of the above message: more delay until we can land this on master and until a new version of this extension can be released for GNOME 40.

**Comment by fin-ger** [comment]
And I think that the frequency configuration gets reset when opening the preferences window of cpupower, but I have to investigate on this.

**Comment by schaze** [comment]
Test on Arch Linux - Gnome 40.2.0

1. Do clean installation - OK
2. Moving both sliders in the dropdown with
  I. Mouse - OK
  II. Arrow keys -OK
  III. Mouse wheel OK - but feels sluggish (using trackpad on laptop)
3. Toggle Turbo Boost - OK -> but menu closes on toggle
- using cpufreq backend I get:
cat: /sys/devices/system/cpu/cpufreq/boost: No such file or directory
/usr/local/bin/cpufreqctl-schaze: line 377: /sys/devices/system/cpu/cpufreq/boost: Permission denied
- all other backends OK
4. Change Profile - OK
5. Toggle Auto-Switch (real testing not possible as VMs don't fire AC adapter events) - OK -> but menu closes on toggle
6. Open Preferences and toggle all options - OK
Remarks:
- on opening the preference window I get the following warning in the console:
(gnome-shell:726112): GnomeDesktop-WARNING **: 20:13:12.366: Could not create transient scope for PID 731909: GDBus.Error:org.freedesktop.DBus.Error.NameHasNoOwner: Name "org.freedesktop.systemd1" does not exist
- When choosing cpufreq I get the following message:
Gjs-Message: 20:22:05.127: JS LOG: [cpupower-status] FrequencyScalingDriver: cpufreq
cat: /sys/devices/system/cpu/cpufreq/boost: No such file or directory
- When choosing intel_pstate I get the following message:
Gjs-Message: 20:23:23.581: JS LOG: [cpupower-status] FrequencyScalingDriver: intel_pstate
/usr/local/bin/cpufreqctl-schaze: line 377: /sys/devices/system/cpu/cpufreq/boost: Permission denied
Gjs-Message: 20:23:23.633: JS LOG: [cpupower-status] Failed to reset frequency scaling driver of old backend cpufreq: UNKNOWN
- when choosing automatic (intel_pstate) no messages are shown

7. Do 2-4 for all available backends - see comments in 2-4
8. Create a new profile and change its position (move up and down in the list)
- when changing frequency for the profile and the moving it before saving, changes are lost
- can NOT confirm that discard button has to be clicked twice  - one click works for me
9. Set all values in that new profile - OK
10. Select it in the extensions dropdown menu - OK
11. Remove the profile - OK
12. Open About Window - OK
13. Uninstall tool and polkit rules
- tool uninstall triggers a lot of log entries - OK -  mostly related to other extensions not sure

**Comment by fin-ger** [comment]
>        III. Mouse wheel OK - but feels sluggish (using trackpad on laptop)
Yes, this is to be expected as scrolling (especially on a laptop) generates a huge amount of events which all set the cpu frequency setting. Setting the CPU frequency configuration setting to often within a short timespan will make your system lag. This might be fixed at a later point by buffering the scroll events to e.g. only update the frequency every 0.5 seconds.


>     * using cpufreq backend I get:
>       cat: /sys/devices/system/cpu/cpufreq/boost: No such file or directory
>       /usr/local/bin/cpufreqctl-schaze: line 377: /sys/devices/system/cpu/cpufreq/boost: Permission denied

Oh! Good catch. It seems like the turbo boost setting is not available for intel CPUs when using the cpufreq driver... This is not really covered by our API. I could make the cpufreq backend unavailable for intel CPUs for now, so that this error does not appear, although the problem is only affecting the turbo boost setting. Alternatively, we could make the turbo boost setting optional for the backends and signal unsupported boost setting for intel CPU's under the cpufreq driver.

>     2. Toggle Auto-Switch (real testing not possible as VMs don't fire AC adapter events) - OK -> but menu closes on toggle
Yes, this is wanted behavior by the GNOME shell widget. We won't change that behavior for consistency with the rest of the GNOME desktop. Although, I agree that toggle buttons should not close the menu... Actually, the default way to toggle these buttons in GNOME is to use the spacebar instead of your left mouse button. Admittedly, I never used the spacebar for the activation of UI controls prior to investigating on this issue.

>     * on opening the preference window I get the following warning in the console:
>       (gnome-shell:726112): GnomeDesktop-WARNING **: 20:13:12.366: Could not create transient scope for PID 731909: GDBus.Error:org.freedesktop.DBus.Error.NameHasNoOwner: Name "org.freedesktop.systemd1" does not exist
Yeah, dbus behaves very weird from inside GNOME shell extensions. I believe this also has something to do with the uninstallation issue.

>     * When choosing cpufreq I get the following message:
>       Gjs-Message: 20:22:05.127: JS LOG: [cpupower-status] FrequencyScalingDriver: cpufreq
>       cat: /sys/devices/system/cpu/cpufreq/boost: No such file or directory
Same issue as above.

>     * When choosing intel_pstate I get the following message:
>       Gjs-Message: 20:23:23.581: JS LOG: [cpupower-status] FrequencyScalingDriver: intel_pstate
>       /usr/local/bin/cpufreqctl-schaze: line 377: /sys/devices/system/cpu/cpufreq/boost: Permission denied
Yes, when changing the backend, the old backend is reset to its default state to prevent interference of several backends.

>       Gjs-Message: 20:23:23.633: JS LOG: [cpupower-status] Failed to reset frequency scaling driver of old backend cpufreq: UNKNOWN
Resetting failed because of cpufreq boost things.

>     * when choosing automatic (intel_pstate) no messages are shown
Perfect!

>     * when changing frequency for the profile and the moving it before saving, changes are lost
Ah thanks! This needs to be fixed!

>     * can NOT confirm that discard button has to be clicked twice  - one click works for me
Yes, sorry I already fixed it but did not update the pull request.

>     * tool uninstall triggers a lot of log entries - OK -  mostly related to other extensions not sure
Yeah, this is dbus issues, polkit updates, etc. Not really perfect but works at least sometimes...

Summary for myself:

 - [ ] cpufreq backend boost on intel CPUs not working
 - [ ] edited profile is lost when moving position
 - [ ] dbus behaves weird, I don't really know what the issue is but on some systems uninstall freezes reproducibly

File: src/indicator.js [hunk 1]
```diff
@@ -29,167 +29,203 @@ const St = imports.gi.St;
...
-                this._updateMin();
-                this._updateMax();
-                this._updateTurbo();
-                this._updateAutoSwitch();
+                this.updateMin();
+                this.updateMax();
+                this.updateTurbo();
+                this.updateAutoSwitch();
             }
         } else {
```
Comment (fin-ger) [inline src/indicator.js:80]: This is not working anymore, most likely it's again another race condition...

What happens:
After login the minimum and maximum frequency are not restored but are always `min: hardware minimum` and `max: 100`, whereas the turbo boost setting is applied from the last session.

What is expected:
Restoring minimum and maximum frequency of last boot just works.

**Comment by fin-ger** [comment]
I hope I now fixed all bugs related to loading the settings of the last session with AutoSwitch, `None` settings in AutoSwitch, plugged in AC on boot, etc. If anyone finds further issues please don't hesitate to report!

**Comment by fin-ger** [comment]
@mastercaution @martin31821 Anyone up for a code review? I'm particularly interested in privilege escalations and security issues. The functionality has been thoroughly tested by other users so far. So there is no need to waste any time in extensive VM testing, etc.

**Review by Jasper-Ben — APPROVED** [review]
LGTM :+1:

**Comment by HenriWahl** [comment]
Is there any way to test this already?

**Comment by fin-ger** [comment]
@HenriWahl Please refer to the `Test Procedure` description at the top of this PR.

**Comment by HenriWahl** [comment]
> @HenriWahl Please refer to the `Test Procedure` description at the top of this PR.
> 
> Edit: And make sure you check out the `feature/modular-cpufreqctl-backends` branch of this PR.

Thanks for the hint, I did miss it.
Sad to say I get an error:

```
version 9.0.4
error: unknown action --backend automatic --format json min get
usage: cpufreqctl [-h] [turbo {get,0,1}] [min {get,check,VALUE}] [max {get,VALUE}]

```

**Comment by fin-ger** [comment]
@HenriWahl See https://github.com/martin31821/cpupower/issues/145#issuecomment-844966360

**Comment by HenriWahl** [comment]
> @HenriWahl See [#145 (comment)](https://github.com/martin31821/cpupower/issues/145#issuecomment-844966360)

Thanks again, now it works!

**Comment by HenriWahl** [comment]
I am just not sure if the profiles should be uneditable or not? Looks like this:
grafik

**Comment by fin-ger** [comment]
@HenriWahl Hmm, this should only happen when you edit a single profile. I guess the profiles tab thinks that you edited the `Quiet` and `Energy Saver` profiles because the minimum is set to `0` which is now impossible due to the profiles manager respecting the hardware limitation of `16%`. Can you try to edit those profiles manually with `dconf` under the path `/org/gnome/shell/extensions/cpupower/profiles` to set the minimum frequency to `16` and see if the issue still appears?

**Comment by HenriWahl** [comment]
Once again: yes, that was the solution. Maybe a good idea to automatically adjust older 0-values to the minimum 16?

**Comment by fin-ger** [comment]
The minimum differs for different CPU models, but yes: updating the profiles to reflect the actual minimum sounds like a good idea :+1:

**Comment by HenriWahl** [comment]
Apparently automatic change when disconecting power cord does not work - won't flip to battery powersave profile.

**Review by martin31821 — APPROVED** [review]

**Comment by fin-ger** [comment]
Alright, so then this is almost ready for merge, expect the last tiny bug @HenriWahl pointed out which causes old profiles to brick the profile editor.

If I missed to fix any of the bugs mentioned in this thread, please point them out now.

**Comment by fin-ger** [comment]
Just tried my changes on an Intel machine and on an AMD machine. Seems to work perfectly, although only Fedora 33 has been tested. As the code changes are quite minimal, I will just assume it works on all the other distros. Will merge it now, however I think the release for a new version should include fixes for #149 and #140. I'm fixing those right now.

**Comment by fin-ger** [comment]
This extension is now available for dozens of Linux distributions (including Gnome 40 support). Consult the readme for detailed instructions on how to install it.

**Pull request merged**